### PR TITLE
[Merged by Bors] - feat(NumberTheory): add converse of Lucas test for primes

### PR DESCRIPTION
**Pull request by Aaron1011**
Add the converse of the existing `lucas_primality` theorem, and combine them into a `lucas_primality_iff` theorem

Coauthored by @vihdzp 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:61]:
```suggestion
```

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:63]: There's a lot of awkward space here, can you reindent the line to fully utilize the 100 characters? (or at least, split it at some point that makes more sense).

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:66]:
```suggestion
such that `a^(p-1) = 1 mod p` and `a^((p-1)/q) ≠ 1 mod p` for all prime factors `q` of `p-1`.
The multiplicative group mod `p` is cyclic, so `a` can be any generator of the group
(which must have order `p-1`).
```

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:131]:
```suggestion
  ⟨reverse_lucas_primality p, fun ⟨a, ⟨ha, hb⟩⟩ ↦ lucas_primality p a ha hb⟩ 
```

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:118]: Here's a condensed version of your proof:
```suggestion
  have : Fact p.Prime := ⟨hP⟩
  obtain ⟨g, hg⟩ := IsCyclic.exists_generator (α := (ZMod p)ˣ)
  have h1 : orderOf g = p - 1 := by
    rwa [orderOf_generator_eq_natCard hg, Nat.card_eq_fintype_card, ← Nat.prime_iff_card_units]
  have h2 := tsub_pos_iff_lt.2 hP.one_lt
  rw [← orderOf_injective (Units.coeHom _) Units.ext _, orderOf_eq_iff h2] at h1
  refine ⟨g, h1.1, fun q hq hqd ↦ ?_⟩
  replace hq := hq.one_lt
  exact h1.2 _ (Nat.div_lt_self h2 hq) (Nat.div_pos (Nat.le_of_dvd h2 hqd) (zero_lt_one.trans hq))
```
I removed the comments, since I don't think they're terribly valuable now that the proof is relatively short (and since the docstring already explains the proof anyways). But feel free to add them back if you think they still provide value.

Comment (vihdzp) [reply Mathlib/NumberTheory/LucasPrimality.lean:118]: Here's some tips that might be helpful in the future:
- `linarith` is a very powerful tactic, and for that reason, it should be reserved for whenever it's absolutely necessary. Almost all uses in your proof were equivalent to `assumption` (and the remaining one was just `1 < q → 0 < q`, so still overkill).
- There usually isn't much reason to use `trivial`, especially when you can replace it by `assumption` or `exact`.
- When you have multiple goals, you should isolate each using `·` and an indent instead of solving them one after the other.

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:81]:
```suggestion

/-- A number `p` is prime if and only if there exists an `a` such that
`a^(p-1) = 1 mod p` and `a^((p-1)/q) ≠ 1 mod p` for all prime factors `q` of `p-1`.
-/
```

**Review by vihdzp — APPROVED** [review]

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (Aaron1011) [reply Mathlib/NumberTheory/LucasPrimality.lean]: Thanks for the help - this is much cleaner than what I came up with. I updated your PR to use your version of the proof, and made another simplification from my original (`orderOf_generator_eq_natCard, Nat.card_eq_fintype_card` can be replaced with `orderOf_eq_card_of_forall_mem_zpowers` ).

Since this is such an extensive simplification, I added you as a co-author

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (vihdzp) [inline Mathlib/NumberTheory/LucasPrimality.lean:79]: Missed a period here too.

**Comment by Aaron1011** [comment]
Are there any other changes that you'd like me to make, or is this ready to merge?

**Comment by vihdzp** [comment]
I'm not an official Mathlib maintainer, so I can't really make that decision. One of them will have to look at the PR and approve it before it can get merged.

**Comment by Aaron1011** [comment]
Oh, my bad. Thanks for the review!

**Comment by b-mehta** [comment]
Looks good to me, but I'll let another maintainer check too.

maintainer merge

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (jcommelin) [inline Mathlib/NumberTheory/LucasPrimality.lean:66]:
```suggestion
theorem reverse_lucas_primality (p : ℕ) (hP : p.Prime) :
```

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/NumberTheory/LucasPrimality.lean
Comment (b-mehta) [reply Mathlib/NumberTheory/LucasPrimality.lean:66]: and similarly on line 81

**Pull request closed without merging**
